### PR TITLE
[FEATURE] Implement Search Functionality for Job Applications

### PR DESCRIPTION
**Pull request by Natata08**
# [FEATURE] Implement Search Functionality for Job Applications

---

## Description

Added search functionality to filter job applications by company name and job title, improving user experience by allowing quick access to specific applications. The implementation includes real-time filtering with a debounced search and an empty state display for no results.

- Problem: Users needed a way to quickly find specific job applications
- Impact: Improved application accessibility and user efficiency in managing multiple job applications

---

## Changes Made

Frontend:
1. Search Implementation:
- Added debounced search using `react-use` library
- Implemented filtering logic for company names and job titles
- Case-insensitive search with whitespace handling
2. Components:
- Modified SearchField component to handle search input
- Added EmptyState component for both no applications and no search results
- Updated ApplicationsBoard to handle filtered results
3. State Management:
- Added search state in DashboardPage
- Implemented filtered applications logic

---

## Screenshots (if applicable)
screencapture-localhost-3000-user-2024-11-13-21_53_43

screencapture-localhost-3000-user-2024-11-13-21_52_57

screencapture-localhost-3000-user-2024-11-13-21_52_36

---

## Testing Steps

1. Search Functionality:
- Enter a company name in the search field
- Enter a job title in the search field
- Enter partial text to verify partial matches
- Clear search field to verify all applications return
2. Empty States:
- Search for non-existent company/title to verify empty state
- Verify empty state message changes based on search context
3. Edge Cases:
- Test with empty string
- Test with whitespace only
- Test with mixed case

---

## Issues Addressed

Closes [Add search functionality](https://trello.com/c/TcwaJYBo)

### CODE REVIEW

**Review by olhasashchuk — DISMISSED** [review]

File: frontend/app/user/page.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 'use client'
 
 import { useState, useEffect } from 'react'
+import { useDebounce } from 'use-debounce'
```
Comment (pankaj28843) [inline frontend/app/user/page.js:4]: I would suggest using this library instead: https://github.com/streamich/react-use

It has `useDebounce` but also many other things, you won't need to add a new package for every new reusable hook you want.

**Review by pankaj28843 — CHANGES_REQUESTED** [review]

**Review by pankaj28843 — APPROVED** [review]

**Review by FurkannOzbek — APPROVED** [review]
